pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -315,23 +324,61 @@ func GetRegistries(ctx *types.SystemContext) ([]Registry, error) {
 
 // FindUnqualifiedSearchRegistries returns all registries that are configured
 // for unqualified image search (i.e., with Registry.Search == true).
-func FindUnqualifiedSearchRegistries(registries []Registry) []Registry {
+func FindUnqualifiedSearchRegistries(ctx *types.SystemContext) ([]Registry, error) {
+	registries, err := GetRegistries(ctx)
+	if err != nil {
+		return nil, err
+	}
+
 	unqualified := []Registry{}
 	for _, reg := range registries {
 		if reg.Search {
 			unqualified = append(unqualified, reg)
 		}
 	}
-	return unqualified
+	return unqualified, nil
+}
+
+// isPrefix checks if the specified prefix is a namespace or an image prefix of
+// reference.  The basic assumption is that the remainder of the reference is
+// either an empty string (i.e., a full match) or a string starting with '/'
+// (i.e., a namespace prefix) or a string starting with a ':' (i.e., an image
+// prefix).
+//
+// "docker.io/library/alpine" is a prefix for "docker.io/library/alpine/path"
+// and for "docker.io/library/alpine:latest" but not for "docker.io/library/alpinenext".
+func isPrefix(reference, prefix string) bool {
+	if strings.HasPrefix(reference, prefix) {
+		remainder := reference[len(prefix):]
+		// full match
+		if len(remainder) == 0 {
+			return true
+		}
+		// The regex expects strings that either start with a '/' or
+		// that start with a ':'.  Strings starting with a ':' have a
+		// subsequent lookahead for ':' and '/' to avoid matching
+		// registries at port boundaries.
+		//
+		// This way, it matches "/foobar", "/foobar:latest" and ":latest"
+		// but it doesn't match ":5000/foobar" or ":latest:foobar".
+		reg := regexp.MustCompile(`(?m)^(\/.*|:.[^:\/]*)$`)
```

Comment:
For now let’s go with `remainder == "" || remainder[0] isOneOf(":/@")` then, but document that the valid `prefix` valuers are _only_
- hostname[:port] as a single unit (not supporting a match of hostname prefix against a hostname:port input)
- hostname[:port]/…/ns
- hostname[:port]/…/repo
- hostname[:port]/…/repo:tag
- hostname[:port]/…/repo@digest


The implementation would match a few more inputs, of which only the hostname vs. hostname:port case would actually be valid inputs, and running multiple registries on a single host seems unlikely enough for now.

Reply:
If we follow where the Moby PR goes, we can just match at paths which makes the logic (and entailing problems) simpler. If the prefix doesn't end with a '/' then we append one. It must be documented but it avoid duplicating the quite hairy parsing.

Reply:
If you mean essentially `reference == prefix || strings.HasPrefix(reference, prefix+"/")`, that would make it impossible to match (e.g. configure mirrors) at the level of an individual repo — and it seems to me that that should be possible.